### PR TITLE
make references search multiple documents #135

### PR DESCRIPTION
**Pull request by lhein**
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

File: src/test/java/com/github/cameltooling/lsp/internal/AbstractCamelLanguageServerTest.java [hunk 1]
```diff
@@ -117,22 +117,39 @@ protected CamelLanguageServer initializeLanguageServer(String text) throws URISy
 
 	protected CamelLanguageServer initializeLanguageServer(String text, String suffixFileName) throws URISyntaxException, InterruptedException, ExecutionException {
 		this.extensionUsed = suffixFileName;
-		InitializeParams params = new InitializeParams();
-		params.setProcessId(new Random().nextInt());
-		params.setRootUri(getTestResource("/workspace/").toURI().toString());
+		initializeLanguageServer(getInitParams());
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/test/java/com/github/cameltooling/lsp/internal/AbstractCamelLanguageServerTest.java]: not sure what you are asking for here.

[the remark this answers is not in the document]
Comment (apupier) [reply src/test/java/com/github/cameltooling/lsp/internal/AbstractCamelLanguageServerTest.java]: factorize code https://github.com/lhein/camel-language-server/pull/3

[the remark this answers is not in the document]
Comment (lhein) [reply src/test/java/com/github/cameltooling/lsp/internal/AbstractCamelLanguageServerTest.java]: merged

File: src/test/java/com/github/cameltooling/lsp/internal/references/ReferencesProcessorTest.java [hunk 2]
```diff
@@ -184,6 +217,44 @@ public void testRetrieveNoReferenceWithDifferentDirectIds() throws Exception {
 		testRetrieveReferences(DIFFERENTIDS, 0, new Position(9, 18));
 	}
 	
+	@Test
+	public void testMultiDocumentReferences() throws Exception {
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/test/java/com/github/cameltooling/lsp/internal/references/ReferencesProcessorTest.java]: split into new test class

File: src/main/java/com/github/cameltooling/lsp/internal/references/ReferencesProcessor.java [hunk 3]
```diff
@@ -85,13 +88,17 @@ private boolean isReference(CamelURIInstance camelURIInstanceToSearchReference,
...
-				allCamelUriInstance.put(new CamelURIInstance(uriToParse, endpoint, textDocumentItem), endpoint);
+		for (TextDocumentItem docItem : textDocumentService.getAllOpenDocuments()) {
+			allEndpoints.addAll(parserXMLFileHelper.getAllEndpoints(docItem));
+			for (Node endpoint : allEndpoints) {
+				String uriToParse = CamelXmlHelper.getSafeAttribute(endpoint, "uri");
+				if (uriToParse != null) {
+					allCamelUriInstance.put(new CamelURIInstance(uriToParse, endpoint, docItem), endpoint);
+				}
 			}
+			allEndpoints.clear();
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/main/java/com/github/cameltooling/lsp/internal/references/ReferencesProcessor.java]: done

File: src/main/java/com/github/cameltooling/lsp/internal/CamelTextDocumentService.java [hunk 4]
```diff
@@ -209,4 +210,8 @@ public void didSave(DidSaveTextDocumentParams params) {
 	public TextDocumentItem getOpenedDocument(String uri) {
 		return openedDocuments.get(uri);
 	}
+	
+	public Collection<TextDocumentItem> getAllOpenDocuments() {
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/main/java/com/github/cameltooling/lsp/internal/CamelTextDocumentService.java]: done

File: pom.xml [hunk 5]
```diff
@@ -55,7 +55,7 @@
 		<lsp4j.version>0.5.0</lsp4j.version>
 		<slf4j.version>1.7.6</slf4j.version>
 		<junit.version>4.12</junit.version>
-		<assertj.version>3.11.1</assertj.version>
+		<assertj.version>3.10.0</assertj.version>
```
[the remark this answers is not in the document]
Comment (lhein) [reply pom.xml]: removed from PR

**Comment by lhein** [comment]
created https://github.com/camel-tooling/camel-language-server/issues/193 for the issue with start and end character location mentioned above. This happens also on master.

**Pull request merged**